File: src/simulation.py
"""Core simulation engine.

simulate_market(p, params, seed) -> dict

Produces time series of prices, fair value, sentiment, demands, and returns
for a given population mix on the 2-simplex.
"""

import numpy as np
from typing import Tuple

from src.agents import ModelParams, prospect_value

# Try to use the Rust engine if available
_USE_RUST = False
try:
    from market_pop_dynamics_rust import simulate_market as _rust_simulate
    _USE_RUST = True
except ImportError:
    pass


def simulate_market(
    p: Tuple[float, float, float],
    params: ModelParams = None,
    seed: int = None,
    force_python: bool = False,
) -> dict:
    """Simulate market dynamics for a given population mix.

    Args:
        p: (p1, p2, p3) population fractions, must sum to 1.
        params: model parameters (defaults to ModelParams()).
        seed: random seed for reproducibility.
        force_python: skip Rust engine even if available.

    Returns:
        dict with time series: prices, fair_value, sentiment, demands, returns, mispricing.
    """
    if params is None:
        params = ModelParams()

    if _USE_RUST and not force_python:
        return _simulate_rust(p, params, seed)
    return _simulate_python(p, params, seed)


def _simulate_python(
    p: Tuple[float, float, float],
    params: ModelParams,
    seed: int = None,
) -> dict:
    """Pure-Python/NumPy simulation engine."""
    if seed is not None:
        np.random.seed(seed)

    p1, p2, p3 = p
    assert abs(p1 + p2 + p3 - 1.0) < 1e-10, f"Population fractions must sum to 1, got {p1 + p2 + p3}"

    n = params.n_steps
    dt = params.dt
    sqrt_dt = np.sqrt(dt)

    # Pre-allocate arrays
    P = np.zeros(n + 1)
    V = np.zeros(n + 1)
    S = np.zeros(n + 1)
    m = np.zeros(n + 1)
    D1 = np.zeros(n + 1)
    D2 = np.zeros(n + 1)
    D3 = np.zeros(n + 1)
    returns = np.zeros(n + 1)

    # Initial conditions
    P[0] = params.P0
    V[0] = params.V_star
    S[0] = 0.0
    m[0] = 0.0

    # Random draws
    dW_f = np.random.randn(n) * sqrt_dt
    dW_s = np.random.randn(n) * sqrt_dt
    dW_n = np.random.randn(n) * sqrt_dt

    decay = params.ewma_decay

    for t in range(n):
        # Fair value: GBM with zero drift
        V[t + 1] = V[t] * np.exp(-0.5 * params.sigma_f**2 * dt + params.sigma_f * dW_f[t])

        # Type I: mean-revert to fair value (normalized by price)
        D1[t] = params.kappa * (V[t] - P[t]) / P[t]

        # Type II: momentum + sentiment
        D2[t] = params.alpha * m[t] + params.beta * S[t]

        # Type III: pure noise
        D3[t] = params.sigma_n * dW_n[t] / sqrt_dt

        # Aggregate demand -> price change
        D_total = p1 * D1[t] + p2 * D2[t] + p3 * D3[t]

        # Price update (log-normal to prevent negative prices)
        dP = params.lam * D_total * dt
        P[t + 1] = P[t] * np.exp(dP - 0.5 * (params.lam * D_total) ** 2 * dt**2)

        # Log return
        returns[t + 1] = np.log(P[t + 1] / P[t])

        # Update momentum (EWMA of returns)
        m[t + 1] = decay * m[t] + (1 - decay) * returns[t + 1]

        # Update sentiment with prospect-theoretic asymmetry
        v_r = prospect_value(
            np.array([returns[t + 1]]),
            lam_pt=params.lam_pt,
            a=params.a_pt,
        )[0]
        dS = -params.gamma * S[t] * dt + params.delta * v_r * dt + params.sigma_s * dW_s[t]
        S[t + 1] = S[t] + dS

    time = np.arange(n + 1) * dt

    return {
        "time": time,
        "price": P,
        "fair_value": V,
        "sentiment": S,
        "momentum": m,
        "demand_value": D1,
        "demand_emotion": D2,
        "demand_noise": D3,
        "returns": returns,
        "mispricing": np.log(P / V),
    }


def _simulate_rust(
    p: Tuple[float, float, float],
    params: ModelParams,
    seed: int = None,
) -> dict:
    """Dispatch to Rust engine and convert result to Python dict."""
    from market_pop_dynamics_rust import simulate_market as rust_sim

    result = rust_sim(
        p[0], p[1], p[2],
        params.T, params.dt, params.V_star, params.P0,
        params.lam, params.sigma_f, params.kappa,
        params.alpha, params.beta, params.tau_m,
        params.gamma, params.delta, params.sigma_s,
        params.lam_pt, params.a_pt, params.sigma_n,
        seed if seed is not None else -1,
    )
    return result

File: rust_core/src/simulation.rs
use rand::prelude::*;
use rand_chacha::ChaCha8Rng;

use crate::agents::prospect_value;
use crate::types::{SimParams, SimResult};

/// Core simulation loop — matches Python's _simulate_python exactly.
pub fn simulate_market(
    p1: f64,
    p2: f64,
    p3: f64,
    params: &SimParams,
    seed: Option<u64>,
) -> SimResult {
    let n = params.n_steps();
    let dt = params.dt;
    let sqrt_dt = dt.sqrt();

    let mut rng: ChaCha8Rng = match seed {
        Some(s) => ChaCha8Rng::seed_from_u64(s),
        None => ChaCha8Rng::from_rng(rand::thread_rng()).unwrap(),
    };

    // Pre-allocate
    let len = n + 1;
    let mut price = vec![0.0f64; len];
    let mut fair_value = vec![0.0f64; len];
    let mut sentiment = vec![0.0f64; len];
    let mut momentum = vec![0.0f64; len];
    let mut d1 = vec![0.0f64; len];
    let mut d2 = vec![0.0f64; len];
    let mut d3 = vec![0.0f64; len];
    let mut returns = vec![0.0f64; len];

    // Initial conditions
    price[0] = params.p0;
    fair_value[0] = params.v_star;

    let decay = params.ewma_decay();

    // Pre-generate random draws (matching Python's randn * sqrt_dt pattern)
    let dw_f: Vec<f64> = (0..n).map(|_| standard_normal(&mut rng) * sqrt_dt).collect();
    let dw_s: Vec<f64> = (0..n).map(|_| standard_normal(&mut rng) * sqrt_dt).collect();
    let dw_n: Vec<f64> = (0..n).map(|_| standard_normal(&mut rng) * sqrt_dt).collect();

    for t in 0..n {
        // Fair value: GBM with zero drift
        fair_value[t + 1] =
            fair_value[t] * (-0.5 * params.sigma_f * params.sigma_f * dt + params.sigma_f * dw_f[t]).exp();

        // Type I: mean-revert to fair value
        d1[t] = params.kappa * (fair_value[t] - price[t]) / price[t];

        // Type II: momentum + sentiment
        d2[t] = params.alpha * momentum[t] + params.beta * sentiment[t];

        // Type III: noise
        d3[t] = params.sigma_n * dw_n[t] / sqrt_dt;

        // Aggregate demand
        let d_total = p1 * d1[t] + p2 * d2[t] + p3 * d3[t];

        // Price update (log-normal)
        let dp = params.lam * d_total * dt;
        price[t + 1] = price[t] * (dp - 0.5 * (params.lam * d_total).powi(2) * dt * dt).exp();

        // Log return
        returns[t + 1] = (price[t + 1] / price[t]).ln();

        // Momentum (EWMA)
        momentum[t + 1] = decay * momentum[t] + (1.0 - decay) * returns[t + 1];

        // Sentiment with prospect theory
        let v_r = prospect_value(returns[t + 1], params.lam_pt, params.a_pt);
        let ds = -params.gamma * sentiment[t] * dt + params.delta * v_r * dt + params.sigma_s * dw_s[t];
        sentiment[t + 1] = sentiment[t] + ds;
    }

    // Compute mispricing: log(P/V)
    let mispricing: Vec<f64> = price
        .iter()
        .zip(fair_value.iter())
        .map(|(p, v)| (p / v).ln())
        .collect();

    // Time array
    let time: Vec<f64> = (0..len).map(|i| i as f64 * dt).collect();

    SimResult {
        time,
        price,
        fair_value,
        sentiment,
        momentum,
        demand_value: d1,
        demand_emotion: d2,
        demand_noise: d3,
        returns,
        mispricing,
    }
}

/// Box-Muller standard normal.
#[inline]
fn standard_normal(rng: &mut ChaCha8Rng) -> f64 {
    use std::f64::consts::TAU;
    let u1: f64 = rng.gen::<f64>();
    let u2: f64 = rng.gen::<f64>();
    (-2.0_f64 * u1.ln()).sqrt() * (TAU * u2).cos()
}
